noresm3_0_022_cam6_4_121: Use CDEPS for CO2 fossil fuel and aircraft emissions

mvertens opened the pull request:
Summary: Enable fossil fuel and aircraft co2 forcing that are currently on a 2 degree FV grid to be remapped online to the model grid and time interpolated to the model time using CDEPS in-line functionality. The grid mapping is currently hard-wired to be first order conservative.

@adagj  @DirkOlivie @tjiputra  - the following lists how you can get and use the new code (still in testing)

to get the current branch:
```
> git clone https://github.com/mvertens/NorESM noresm3_0_beta03b_co2refactor
> cd noresm3_0_beta03b_co2refactor
> git checkout feature/refactor_co2input 
> ./bin/git-fleximod update
```

compset and resolution
```
--compset 1850_CAM70%LT%NORESM%CAMoslo_CLM60%FATES_CICE_BLOM%HYB%ECO_MOSART_DGLC%NOEVOLVE_SWAV_SESP_BGC%BPRP
-res ne16pg3_tn14
```

user_nl_cam 

- the key additions are in the namelist group `co2flux_fuel_mesh, co2_flux_fuel_year_first, co2_flux_fuel_year_last, co2_flux_fuel_year_align`


```
interpolate_nlat           = 96
interpolate_nlon           = 144
interpolate_output         = .true.
use_aerocom                = .false.
history_aerosol            = .false.
zmconv_c0_lnd              =  0.0075D0
zmconv_c0_ocn              =  0.0300D0
zmconv_ke                  =  5.0E-6
zmconv_ke_lnd              =  1.0E-5
clim_modal_aero_top_press  =  1.D-4
bndtvg                     = '/cluster/shared/noresm/inputdata/atm/cam/ggas/noaamisc.r8.nc'
dust_emis_method           = 'Leung_2023'
dust_emis_fact             = 6.1D0
rafsip_on                  = .true.
micro_mg_dcs               = 700.D-6
clubb_c8                   =  5.3
clubb_l_do_expldiff_rtm_thlm = .true.
!
co2_flag                = .true.
!
co2_readflux_aircraft   = .true.
aircraft_co2_datafile   = '/cluster/shared/noresm/inputdata/atm/cam/ggas/emissions-cmip6_CO2_anthro_ac_ScenarioMIP_IAMC-REMIND-MAGPIE-ssp534_201401-210112_fv_1.9x2.5_c20220809.nc'
aircraft_co2_fldname    = 'ac_CO2'
aircraft_co2_meshfile   = '/cluster/shared/noresm/inputdata/share/meshes/fv1.9x2.5_141008_ESMFmesh_c20191001.nc'
aircraft_co2_year_first = 2014
aircraft_co2_year_last  = 2101
aircraft_co2_year_align = 2014
aircraft_co2_tintaglo = 'linear'
aircraft_co2_taxmode    = 'limit'
!
co2_readflux_fuel       = .true.
!
co2flux_fuel_datafile   = '/cluster/projects/nn9560k/adagj/co2files/emissions-cmip6_CO2_anthro_surface_2xScenarioMIP_IAMC-REMIND-MAGPIE-ssp534_201401-210112_fv_1.9x2.5_c20220809.nc'
co2flux_fuel_meshfile   = '/cluster/shared/noresm/inputdata/share/meshes/fv1.9x2.5_141008_ESMFmesh_c20191001.nc'
co2flux_fuel_year_first = 2014
co2flux_fuel_year_last  = 2101
co2flux_fuel_year_align = 2014
co2flux_fuel_tintalgo = 'linear'
co2flux_fuel_taxmode    = 'limit'
```

Reviewers: @gold2718 

Changes made to build system: None

Changes made to the namelist: The following are new namelist variables:
New namelist` co2_ffuel_nl:`
```
- co2flux_fuel_datafile, co2flux_fuel_meshfile, 
- co2flux_fuel_year_first, co2flux_fuel_year_last, co2flux_fuel_year_align
- co2flux_fuel_tintalgo, co2flux_fuel_taxmode
```
 
New namelist `aircraft_emit_nl`:
```
- aircraft_co2_datafile,  aircraft_co2_meshfile, 
- aircraft_co2_year_first,  aircraft_co2_year_last, aircraft_co2_year_align 
- aircraft_co2_tintalgo, aircraft_co2_taxmode
- aircraft_h2o_datafile, aircraft_h2o_meshfile,
- aircraft_h2o_year_first, aircraft_h2o_year_last, aircraft_h2o_year_align
- aircraft_h2o_tintalgo, aircraft_h2o_taxmode
- aircraft_slant_dist_datafile, aircraft_slant_dist_meshfile,
- aircraft_slant_dist_year_first, aircraft_slant_dist_year_last, aircraft_slant_dist_year_align
- aircraft_slant_dist_tintalgo, aircraft_slant_dist_taxmode
```
- to understand these definitions in more detail see 
https://escomp.github.io/CDEPS/versions/master/html/index.html
https://escomp.github.io/CDEPS/versions/master/html/streams.html#data-model-stream-inline-api (this is the functionality used in co2_cycle.F90 and co2_data_flux.F90)

Changes to the defaults for the boundary datasets: None

Substantial timing or memory changes: [ describe or write 'None' ]

Validation:
- The left plot is the model output on day 16 (2 degree forcing mapped to ne16 and then mapped back out to 2 degree history output) 
- The right plot is the forcing dataset /cluster/shared/noresm/inputdata/atm/cam/ggas/emissions-cmip6_CO2_anthro_surface_ScenarioMIP_IAMC-REMIND-MAGPIE-ssp534_201401-210112_fv_1.9x2.5_c20220809.nc on January 16
- 
<img width="1578" height="1033" alt="Screenshot 2025-11-04 at 3 43 40 PM" src="https://github.com/user-attachments/assets/2a124e41-e8d6-4493-b1fb-ce6c1b704edc" />

Testing:
- BFB: Tested `aux_cam_noresm` against NorESM 3.0 beta09 baseline with only namelist changes due to always having the `co2_cycle_nl` namelist in `atm_in`.

Commit:
first pass at refactoring the nudging code

mvertens committed:
fixed compiler issues

Commit:
more refactor of code

mvertens committed:
migrated atm_stream_nuding.F90 code into nudging.F90 and added atm_shr.F90 for model_mesh and model_clock

mvertens committed:
fixes to nudging for running

Commit:
refactor of co2_cycle and co2_data_flux to use cdeps streams

mvertens committed:
first pass at the validation

tjiputra commented:
@mvertens, this command doesn't work
./bin/git-fleximod update

mvertens commented:
> @mvertens, this command doesn't work ./bin/git-fleximod update

@tjiputra - sorry about that - my checkout missed that you need to invoke
`>  git checkout feature/refactor_co2input `
I've updated the documentation.

adagj commented:
@mvertens Thank you Mariana! I will run a N1850 compset with 2xSSP534 CO2 flux data to see if we can track down the (potential) CO2 conservation issue in CAM

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
+Time interpolation algorithm to use for co2flux_ocn_file.
+Default: linear
+</entry>
+<entry id="co2flux_ocn_tintalgo" type="integer" category="co2_cycle"
```
This has the same name as the previous entry. Supposed to be something else?
```suggestion
<entry id="co2flux_ocn_taxmode" type="integer" category="co2_cycle"
```

mvertens replied:
Done.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -220,32 +220,30 @@
        group="nudging_nl" valid_values="" >
-       Full pathname of analyses data to use for nudging.
-        (e.g. '/$DIN_LOC_ROOT/atm/cam/nudging/')
+       Full pathnames of analyses data to use for nudging.
```
I think this should be:
```suggestion
       List of the absolute paths of the filenames of analyses data to use for nudging.
```

mvertens replied:
I have backed out the nudging.F90 code and associated namelist - so I will take care of this in the nudging PR.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -220,32 +220,30 @@
        </entry>
 
-<entry id="Nudge_Path" type="char*256" input_pathname="abs" category="nudging"
+<entry id="Nudge_Filenames" type="char*256(1000)" category="nudging"
```
Should this also include:
```
input_pathname="abs"
```
?

mvertens replied:
Nudging is no longer part of this PR - so the above is no longer relevant. I have put this change in the Nudging PR #243

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -220,32 +220,30 @@
        group="nudging_nl" valid_values="" >
-       Template for Nudging analyses file names.
-        (e.g. '%y/ERAI_ne30np4_L30.cam2.i.%y-%m-%d-%s.nc')
+       ESMF mesh file that corresponds to the nudging files
```
Is this also a full pathname?
```suggestion
       Full pathname to ESMF mesh file that corresponds to the nudging files
```

mvertens replied:
This will be addressed in the nudging PR.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -220,32 +220,30 @@
        </entry>
 
-<entry id="Nudge_File_Template" type="char*80" category="nudging"
+<entry id="Nudge_Meshfile" type="char*256" category="nudging"
```
Should this also include:
``` 
input_pathname="abs"
```
?

mvertens replied:
This is gone in the nudging PR. Unfortunately the nudging.F90 file came in here but without the associated changes to the namelist. This is gone in the nudging PR.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -220,32 +220,30 @@
        </entry>
 
-<entry id="Nudge_Times_Per_Day" type="integer" category="nudging"
+<entry id="Nudge_File_Times_Per_Day" type="integer" category="nudging"
```
Can this (at least in principle) be determined from the file itself? This feels prone to error (i.e., what happens if you specify the wrong value?).

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
 Default: none
 </entry>
+<entry id="co2flux_ocn_mesh" type="char*256" input_pathname="abs" category="co2_cycle"
+       group="co2_cycle_nl" valid_values="" >
```
Should this also include:
```
input_pathname="abs"
```
?

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
 </entry>
+<entry id="co2flux_ocn_mesh" type="char*256" input_pathname="abs" category="co2_cycle"
+       group="co2_cycle_nl" valid_values="" >
+Filepath for ESMF Mesh corresponding to co2flux_ocn_file File.
```
Is this a full pathname?
```suggestion
Full filepath for ESMF Mesh corresponding to co2flux_ocn_file File.
```

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
+</entry>
+<entry id="co2flux_ocn_year_first" type="integer" category="co2_cycle"
+       group="co2_cycle_nl" valid_values="" >
+Year first to use in co2flux_ocn_file.
```
Is this
```suggestion
First year to use in co2flux_ocn_file.
```

mvertens replied:
I have removed the co2_flux_ocn capability from CAM as part of this PR.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
+Default: 0
+<entry id="co2flux_ocn_year_last" type="integer" category="co2_cycle"
+       group="co2_cycle_nl" valid_values="" >
+Year last to use in co2flux_ocn_file.
```
Is this
```suggestion
Last year to use in co2flux_ocn_file.
```

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
 </entry>
+<entry id="co2flux_fuel_mesh" type="char*256" input_pathname="abs" category="co2_cycle"
+       group="co2_cycle_nl" valid_values="" >
+Filepath for ESMF Mesh corresponding to co2flux_fuel_file File.
```
```suggestion
Full filepath for ESMF Mesh corresponding to co2flux_fuel_file File.
```

mvertens replied:
Dine,

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
+</entry>
+<entry id="co2flux_fuel_year_first" type="integer" category="co2_cycle"
+       group="co2_cycle_nl" valid_values="" >
+Year first to use in co2flux_fuel_file.
```
```suggestion
First year to use in co2flux_fuel_file.
```

mvertens replied:
Done.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
+Default: 0
+<entry id="co2flux_fuel_year_last" type="integer" category="co2_cycle"
+       group="co2_cycle_nl" valid_values="" >
+Year last to use in co2flux_fuel_file.
```
```suggestion
Last year to use in co2flux_fuel_file.
```

mvertens replied:
Done.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
+</entry>
+<entry id="co2flux_fuel_year_align" type="integer" category="co2_cycle"
+       group="co2_cycle_nl" valid_values="" >
+Year align to use in co2flux_fuel_file.
```
I know this is a standard term but a one-sentence explanation of how to use it (and/or what it does) would be helpful.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1801,12 +1799,70 @@ Default: set by build-namelist
+Time interpolation algorithm to use for co2flux_fuel_file.
+Default: linear
+</entry>
+<entry id="co2flux_fuel_tintalgo" type="integer" category="co2_cycle"
```
This has the same name as the previous entry. Supposed to be something else?
```suggestion
<entry id="co2flux_fuel_taxmode" type="integer" category="co2_cycle"
```

mvertens replied:
Done. Thanks for checking this.

gold2718 commented on src/cpl/nuopc/atm_shr.F90:
```diff
@@ -0,0 +1,11 @@
+  public
+
+  type(ESMF_Mesh)  :: model_mesh     ! model mesh
+  type(ESMF_Clock) :: model_clock    ! model clock
```
I think this should be private with protected variables initialized by a subroutine:
```suggestion
  private

  type(ESMF_Mesh), public, protected  :: model_mesh     ! model mesh
  type(ESMF_Clock), public, protected :: model_clock    ! model clock
  
  public :: initialize_model_clock
  public :: initialize_model_mesh
  
  logical :: model_clock_initialized = .false.
  logical :: model_mesh_initialized = .false.
contains

  subroutine initialize_model_clock(mclock)
    use shr_sys_mod, only: shr_sys_abort
    type(ESMF_Clock), intent(in) :: mclock
    
    if (model_clock_initialized) then
      call shr_sys_abort('initialize_model_clock: model clock already initialized')
    else
      model_clock = mclock
      model_clock_initialized = .true.
    end if
  end subroutine initialize_model_clock

  subroutine initialize_model_mesh(mmesh)
    use shr_sys_mod, only: shr_sys_abort
    type(ESMF_Mesh), intent(in) :: mmesh
    
    if (model_mesh_initialized) then
      call shr_sys_abort('initialize_model_mesh: model mesh already initialized')
    else
      model_mesh = mmesh
      model_mesh_initialized = .true.
    end if
  end subroutine initialize_model_mesh
```

mvertens replied:
Done

gold2718 commented on src/cpl/nuopc/atm_comp_nuopc.F90:
```diff
@@ -761,11 +760,12 @@ subroutine InitializeRealize(gcomp, importState, exportState, clock, rc)
-       model_clock = clock
+       ! Create model_clock as a variable in atm_shr.F90 - needed for generating streams
+       model_clock = ESMF_ClockCreate(clock, rc=rc)
+       if (ChkErr(rc,__LINE__,u_FILE_u)) return
```
I think some or all of this should move into atm_shr.F90

mvertens replied:
Done

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -30,30 +30,39 @@ module co2_cycle
+   integer            :: co2flux_ocn_year_last  = -999    ! last year in stream to use
+   integer            :: co2flux_ocn_year_align = -999    ! align stream_year_first
+   character(len=cs)  :: co2flux_ocn_tintalgo = 'linear'  ! time interpolation [lower, upper, nearest, linear or coszen]
+   character(len=cs)  :: co2flux_ocn_taxmode = 'extend'   ! time extraploation [cycle, extend or limit]
```
I think that these should also be initialized to invalid values. This would have caught the missing namelist variables.
```suggestion
   character(len=cs)  :: co2flux_ocn_tintalgo = 'NOTSET'  ! time interpolation [lower, upper, nearest, linear or coszen]
   character(len=cs)  :: co2flux_ocn_taxmode = 'NOTSET'   ! time extraploation [cycle, extend or limit]
```

mvertens replied:
Done

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -30,30 +30,39 @@ module co2_cycle
+   integer            :: co2flux_fuel_year_last = -999    ! last year in stream to use
+   integer            :: co2flux_fuel_year_align = -999   ! align stream_year_first
+   character(len=cs)  :: co2flux_fuel_tintalgo = 'linear' ! time interpolation [lower, upper, nearest, linear or coszen]
+   character(len=cs)  :: co2flux_fuel_taxmode = 'extend'  ! time extraploation [cycle, extend or limit]
```
I think that these should also be initialized to invalid values. This would have caught the missing namelist variables.
```suggestion
   character(len=cs)  :: co2flux_fuel_tintalgo = 'NOTSET' ! time interpolation [lower, upper, nearest, linear or coszen]
   character(len=cs)  :: co2flux_fuel_taxmode = 'NOTSET'  ! time extraploation [cycle, extend or limit]
```

mvertens replied:
Done

gold2718 reviewed (changes requested):
I still have most of co2_cycle.F90 and nudging.F90 to review but here are some items I think should be changed.

mvertens committed:
addressed issues in the PR

mvertens committed:
more updates

mvertens committed:
added aircraft emissions to this PR and also addressed more issues raised in the PR review

mvertens committed:
fixed compilation problem

mvertens committed:
updates for default namelists

mvertens committed:
removed co2_readFlux_ocn

mvertens commented:
@gold2718 - I have removed nudging capability as part of this PR. I have also removed co2_readFlux_ocn from CAM as part of this PR. I think all of the comments above that refer to either of these changes are addressed as part of this removal.

mvertens commented:
@gold2718 - what tests do you want me to run?

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+   type(forcing_type) :: forcing(N_AERO)
+
+   real(r8), parameter :: molmass(N_AERO) = 1._r8
+   character(*),parameter :: u_FILE_u = __FILE__
```
This should be:
```suggestion
   character(len=*),parameter :: u_FILE_u = __FILE__
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+      if (masterproc) then
+         open( newunit=unitn, file=trim(nlfile), status='old' )
+         call find_group_name(unitn, 'aircraft_emit_nl', status=ierr)
+         if (ierr == 0) then
```
There is no `else` here so this namelist is optional? Could you could skip all the processing below if it is not present?

mvertens replied:
@gold2718 - I'm not sure what you mean by this.

gold2718 replied:
Most namelist variables are set to invalid values. If you do not read any namelist values, should the model keep running or is an `else` statement that calls `endrun` needed? After all the refactoring, it may be fine to keep running in which case the code does not need a change. If you agree, please comment and resolve this conversation. Otherwise, please call `endrun` if the namelist group is not found.

Another thought is that if `co2_readflux_aircraft` is `.false.`, maybe you should just `return` before even looking for the namelist?

mvertens replied:
@gold2718 - you still need to read the namelist for h2o and slant_dist aircraft emissions - so I think the code needs to stay the way it is.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+
+         if (forcing(nf)%fldname /= ' ') then
+            call mpi_bcast(forcing(nf)%datafile, len(forcing(nf)%datafile), mpi_character, masterprocid, mpicom, ierr)
+            if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: forcing(nf)%datapath")
```
These messages don't tell you which field-type failed. Could you use this instead (subst for `nf`)?
```suggestion
            if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: forcing("//int2str(nf)//")%datapath")
```
You will need `use string_utils, only: int2str``.

mvertens replied:
Good suggestion. Done.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+               call endrun('aircraft_emit_register: '//trim(forcing(nf)%fldname)//&
+                    ' is not a supported aircraft emission field name')
+            endif
+            forcing(nf)%index_map = index
```
I think this section should be at the top of the `do` loop (currently line 139). Or maybe, the `fldname` should be checked up where you set it just after namelist read (see comments there)? Because of the way these fields are being set starting at line 117, can the last line be anything other than:
```
forcing(nf)%index_map = nf
```

mvertens replied:
@gold2718 - can we please chat about this.

mvertens replied:
Good point. I cleaned this up and also removed more unused variables.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+               if (masterproc) write(iulog,*) ' '
+               write(iulog,*) ' '
+               write(iulog,'(a)'  ) ' aircraft init settings for: '//trim(forcing(nf)%fldname)
+               write(iulog,'(a,a)') '   aircraft datafile   = ',trim(forcing(nf)%datafile)
```
FYI, this can also be written as:
```suggestion
               write(iulog,'(2a)') '   aircraft datafile   = ',trim(forcing(nf)%datafile)
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
          end if
+         close(unitn)
+
+         forcing(1)%fldname    = aircraft_co2_fldname
```
Is anything other than `ac_CO2` acceptable here? Why not check that here? Is case important?

mvertens replied:
This is no longer in the namelist - it is hardcoded to  
   character(len=cs) :: aircraft_co2_fldname  = 'ac_CO2'
and 
  forcing(1)%fldname    = aircraft_co2_fldname

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+         forcing(1)%year_last  = aircraft_co2_year_last
+         forcing(1)%year_align = aircraft_co2_year_align
+
+         forcing(2)%fldname    = aircraft_h2o_fldname
```
Is anything other than ac_H2O acceptable here? Why not check that here? Is case important?

mvertens replied:
This is no longer int he namelist. It is hard-coded to 
character(len=cs) :: aircraft_h2o_fldname          = 'ac_H2O'

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+         forcing(2)%year_last  = aircraft_h2o_year_last
+         forcing(2)%year_align = aircraft_h2o_year_align
+
+         forcing(3)%fldname    = aircraft_slant_dist_fldname
```
Is anything other than ac_SLANT_DIST acceptable here? Why not check that here? Is case important?

mvertens replied:
This is no longer in the namelist - it is hard coded to 
character(len=cs) :: aircraft_slant_dist_fldname   = 'ac_SLANT_DIST'

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+               write(iulog,'(a,a)') '   aircraft datafile   = ',trim(forcing(nf)%datafile)
+               write(iulog,'(a,a)') '   aircraft meshfile   = ',trim(forcing(nf)%meshfile)
+               write(iulog,'(a,a)') '   aircraft mapalgo    = ',trim(forcing(nf)%mapalgo)
+               write(iulog,'(a,a)') '   aircraft tintalgo   = ',trim(forcing(nf)%tintalgo)
```
Should `taxmode` also be output?

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+               write(iulog,'(a,a)') '   aircraft datafile   = ',trim(forcing(nf)%datafile)
+               write(iulog,'(a,a)') '   aircraft meshfile   = ',trim(forcing(nf)%meshfile)
+               write(iulog,'(a,a)') '   aircraft mapalgo    = ',trim(forcing(nf)%mapalgo)
+               write(iulog,'(a,a)') '   aircraft tintalgo   = ',trim(forcing(nf)%tintalgo)
```
Please add back `tintalgo`:
```suggestion
               write(iulog,'(a,a)') '   aircraft tintalgo   = ',trim(forcing(nf)%tintalgo)
               write(iulog,'(a,a)') '   aircraft taxmode   = ',trim(forcing(nf)%taxmode)
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+      !--------------------------------------------
+
+      do nf = 1,N_AERO
+         if (forcing(nf)%fldname /= ' ') then
```
This is more robust as:
```suggestion
         if (len_trim(forcing(nf)%fldname) > 0) then
```
or you could use `NOTSET` instead of an empty string as the module default value for filenames and other strings.

mvertens replied:
The check is no longer on the fldname - but rather on the filename -  
       ` if (trim(forcing(nf)%datafile) /= 'unset') then
`

gold2718 replied:
This is great. Since this also controls the log output, could you add in an `else` clause (i.e., just before the line with `end if datafile_isnot_unset`) to say that this forcing is not available? Something like:
```
         else
            write(iulog, '(2a)') 'No forcing data for ',trim(forcing(nf)%fldname)
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+
+            ! Open file
+            if (masterproc) then
+            end if
```
This section looks like it could be removed.

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+   end subroutine get_aircraft
+
+   !=========================================================================
+   subroutine get_vertical_dimension( fid, dname, dsize, data )
```
This should be declared as a private module function (`private :: get_vertical_dimension`) in the module preamble.

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+
+      call pio_seterrorhandling( fid, PIO_BCAST_ERROR, oldmethod=err_handling)
+      ierr = pio_inq_dimid( fid, dname, id )
+      call pio_seterrorhandling( fid, err_handling)
```
1. Why not set this back at the end of the routine?
2. If you are going to use `PIO_BCAST_ERROR`, I think you should check the return code of each call to see if it was an error.

mvertens replied:
Done.

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+            ! Add field to cam history output
+            call addfld(trim(forcing(nf)%fldname), (/ 'lev' /), 'A', trim(forcing(nf)%fldunits), &
+                 'aircraft emission '//trim(forcing(nf)%fldname))
+            call phys_getopts(history_chemistry_out=history_chemistry)
```
This should be moved to before the loop.

mvertens replied:
Moved call to phys_getopts before the loop - but not the call to addfld since it depends on the loop index.

gold2718 replied:
Right. I'm not sure why those other lines got caught up in this comment.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,632 @@
+            end if
+
+            ! Get index in pbuf
+            forcing(nf)%pbuf_index = pbuf_get_index(forcing(nf)%fldname)
```
This is already set in the `pbuf_add_field` call so it is not needed here.

mvertens replied:
done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -397,13 +397,13 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
        allocate( file%hyam(file%nlev),  file%hybm(file%nlev), stat=astat )
        if( astat /= 0 ) then
-          write(iulog,*) 'trcdata_init: file%hyam,file%hybm allocation error = ',astat
+          write(iulog,'(a,i8)') 'trcdata_init: file%hyam,file%hybm allocation error = ',astat
```
Do you need `i8` here? `i0` always gets you the correct size:
```suggestion
          write(iulog,'(a,i0)') 'trcdata_init: file%hyam,file%hybm allocation error = ',astat
```

mvertens replied:
Good point. Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -397,13 +397,13 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
        if( astat /= 0 ) then
-          write(iulog,*) 'trcdata_init: file%hyam,file%hybm allocation error = ',astat
+          write(iulog,'(a,i8)') 'trcdata_init: file%hyam,file%hybm allocation error = ',astat
           call endrun('trcdata_init: failed to allocate file%hyam and file%hybm arrays')
```
This is another case where using `int2str` (`use string_utils, only: int2str`) can save you having to create a separate `write` statement:
```suggestion
          call endrun('trcdata_init: failed to allocate file%hyam and file%hybm arrays, error code = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -397,13 +397,13 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
        if( astat /= 0 ) then
-          write(iulog,*) 'trcdata_init: file%hyai,file%hybi allocation error = ',astat
+          write(iulog,'(a,i8)') 'trcdata_init: file%hyai,file%hybi allocation error = ',astat
           call endrun('trcdata_init: failed to allocate file%hyai and file%hybi arrays')
```
Comments above about `i0` and/or just outputting the error code in the `endrun` call apply here as well.

mvertens replied:
Outputting all of the error code in endrun throughout tracer_data.F90

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -429,23 +429,23 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
        if( astat/= 0 ) then
-          write(iulog,*) 'trcdata_init: failed to allocate file%ps_in(1)%data array; error = ',astat
+          write(iulog,'(a,i8)') 'trcdata_init: failed to allocate file%ps_in(1)%data array; error = ',astat
           call endrun
```
Naked `endrun` calls are not useful (as the write statement is often lost):
```suggestion
          call endrun('trcdata_init: failed to allocate file%ps_in(1)%data array; error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -429,23 +429,23 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
        if( astat/= 0 ) then
-          write(iulog,*) 'trcdata_init: failed to allocate file%ps_in(2)%data array; error = ',astat
+          write(iulog,'(a,i8)') 'trcdata_init: failed to allocate file%ps_in(2)%data array; error = ',astat
           call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
          call endrun('trcdata_init: failed to allocate file%ps_in(2)%data array; error = '//int2str(astat))
```

mvertens replied:
All endrun calls in tracer_data.F90 now have the appropriate failure statements.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -429,23 +429,23 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
           if( astat/= 0 ) then
-             write(iulog,*) 'trcdata_init: failed to allocate file%ps_in(3)%data array; error = ',astat
+             write(iulog,'(a,i8)') 'trcdata_init: failed to allocate file%ps_in(3)%data array; error = ',astat
              call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
             call endrun('trcdata_init: failed to allocate file%ps_in(3)%data array; error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -429,23 +429,23 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
           if( astat/= 0 ) then
-             write(iulog,*) 'trcdata_init: failed to allocate file%ps_in(4)%data array; error = ',astat
+             write(iulog,'(a,i8)') 'trcdata_init: failed to allocate file%ps_in(4)%data array; error = ',astat
              call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
             call endrun('trcdata_init: failed to allocate file%ps_in(4)%data array; error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -482,7 +482,7 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
           if( astat/= 0 ) then
-             write(iulog,*) 'trcdata_init: failed to allocate flds(f)%data array; error = ',astat
+             write(iulog,'(a,i8)') 'trcdata_init: failed to allocate flds(f)%data array; error = ',astat
              call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
             call endrun('trcdata_init: failed to allocate flds(f)%data array; error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -495,7 +495,7 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
        if( astat/= 0 ) then
-          write(iulog,*) 'trcdata_init: failed to allocate flds(f)%input(1)%data array; error = ',astat
+          write(iulog,'(a,i8)') 'trcdata_init: failed to allocate flds(f)%input(1)%data array; error = ',astat
           call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
          call endrun('trcdata_init: failed to allocate flds(f)%input(1)%data array; error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -504,7 +504,7 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
        if( astat/= 0 ) then
-          write(iulog,*) 'trcdata_init: failed to allocate flds(f)%input(2)%data array; error = ',astat
+          write(iulog,'(a,i8)') 'trcdata_init: failed to allocate flds(f)%input(2)%data array; error = ',astat
           call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
          call endrun('trcdata_init: failed to allocate flds(f)%input(2)%data array; error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -515,7 +515,7 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
           if( astat/= 0 ) then
-             write(iulog,*) 'trcdata_init: failed to allocate flds(f)%input(3)%data array; error = ',astat
+             write(iulog,'(a,i8)') 'trcdata_init: failed to allocate flds(f)%input(3)%data array; error = ',astat
              call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
             call endrun('trcdata_init: failed to allocate flds(f)%input(3)%data array; error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -524,7 +524,7 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
           if( astat/= 0 ) then
-             write(iulog,*) 'trcdata_init: failed to allocate flds(f)%input(4)%data array; error = ',astat
+             write(iulog,'(a,i8)') 'trcdata_init: failed to allocate flds(f)%input(4)%data array; error = ',astat
              call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
             call endrun('trcdata_init: failed to allocate flds(f)%input(4)%data array; error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -628,32 +628,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
         if( astat /= 0 ) then
-           write(iulog,*) 'trcdata_init: file%weight_x allocation error = ',astat
+           write(iulog,'(a,i8)') 'trcdata_init: file%weight_x allocation error = ',astat
            call endrun('trcdata_init: failed to allocate weight_x array')
```
These can be combined:
```suggestion
           call endrun('trcdata_init: failed to allocate file%weight_x array, allocation error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -628,32 +628,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
         if( astat /= 0 ) then
-           write(iulog,*) 'trcdata_init: file%weight_y allocation error = ',astat
+           write(iulog,'(a,i8)') 'trcdata_init: file%weight_y allocation error = ',astat
            call endrun('trcdata_init: failed to allocate weight_y array')
```
These can be combined:
```suggestion
           call endrun('trcdata_init: failed to allocate file%weight_y array, allocation error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -628,32 +628,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
         if( astat /= 0 ) then
-           write(iulog,*) 'trcdata_init: file%count_x allocation error = ',astat
+           write(iulog,'(a,i8)') 'trcdata_init: file%count_x allocation error = ',astat
            call endrun('trcdata_init: failed to allocate count_x array')
```
These can be combined:
```suggestion
           call endrun('trcdata_init: failed to allocate file%count_x array, allocation error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -666,32 +666,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
            allocate(file%index0_x(plon,file%nlon), stat=astat)
-           if( astat /= 0 ) then
-              write(iulog,*) 'trcdata_init: file%index0_x allocation error = ',astat
+           if( astat /= '(a,i8)' ) then
```
Too much cut & paste?
```suggestion
           if( astat /= 0 ) then
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -628,32 +628,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
         if( astat /= 0 ) then
-           write(iulog,*) 'trcdata_init: file%count_y allocation error = ',astat
+           write(iulog,'(a,i8)') 'trcdata_init: file%count_y allocation error = ',astat
            call endrun('trcdata_init: failed to allocate count_y array')
```
These can be combined:
```suggestion
           call endrun('trcdata_init: failed to allocate file%count_y array, allocation error = '//int2str(astat))
```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -628,32 +628,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
         if( astat /= 0 ) then
-           write(iulog,*) 'trcdata_init: file%index_x allocation error = ',astat
+           write(iulog,'(a,i8)') 'trcdata_init: file%index_x allocation error = ',astat
            call endrun('trcdata_init: failed to allocate index_x array')
```
These can be combined:
```suggestion
           call endrun('trcdata_init: failed to allocate file%index_x array, allocation error = '//int2str(astat))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -628,32 +628,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
         if( astat /= 0 ) then
-           write(iulog,*) 'trcdata_init: file%index_y allocation error = ',astat
+           write(iulog,'(a,i8)') 'trcdata_init: file%index_y allocation error = ',astat
            call endrun('trcdata_init: failed to allocate index_y array')
```
These can be combined:
```suggestion
           call endrun('trcdata_init: failed to allocate file%index_y array, allocation error = '//int2str(astat))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -666,32 +666,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
            if( astat /= 0 ) then
-              write(iulog,*) 'trcdata_init: file%weight0_x allocation error = ',astat
+              write(iulog,'(a,i8)') 'trcdata_init: file%weight0_x allocation error = ',astat
               call endrun('trcdata_init: failed to allocate weight0_x array')
```
These can be combined:
```suggestion
              call endrun('trcdata_init: failed to allocate file%weight0_x array, allocation error = '//int2str(astat))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -666,32 +666,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
            if( astat /= 0 ) then
-              write(iulog,*) 'trcdata_init: file%weight0_y allocation error = ',astat
+              write(iulog,'(a,i8)') 'trcdata_init: file%weight0_y allocation error = ',astat
               call endrun('trcdata_init: failed to allocate weight0_y array')
```
These can be combined:
```suggestion
              call endrun('trcdata_init: failed to allocate file%weight0_y array, allocation error = '//int2str(astat))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -666,32 +666,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
            if( astat /= 0 ) then
-              write(iulog,*) 'trcdata_init: file%count0_x allocation error = ',astat
+              write(iulog,'(a,i8)') 'trcdata_init: file%count0_x allocation error = ',astat
               call endrun('trcdata_init: failed to allocate count0_x array')
```
These can be combined:
```suggestion
              call endrun('trcdata_init: failed to allocate file%count0_x array, allocation error = '//int2str(astat))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -666,32 +666,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
            if( astat /= 0 ) then
-              write(iulog,*) 'trcdata_init: file%count0_y allocation error = ',astat
+              write(iulog,'(a,i8)') 'trcdata_init: file%count0_y allocation error = ',astat
               call endrun('trcdata_init: failed to allocate count0_y array')
```
These can be combined:
```suggestion
              call endrun('trcdata_init: failed to allocate file%count0_y array, allocation error = '//int2str(astat))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -666,32 +666,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
-              write(iulog,*) 'trcdata_init: file%index0_x allocation error = ',astat
+           if( astat /= '(a,i8)' ) then
+              write(iulog,'(a,i8)') 'trcdata_init: file%index0_x allocation error = ',astat
               call endrun('trcdata_init: failed to allocate index0_x array')
```
These can be combined:
```suggestion
              call endrun('trcdata_init: failed to allocate file%index0_x array, allocation error = '//int2str(astat))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1034,7 +1034,7 @@ function incr_filename( filename, filenames_list, datapath, cyclical_list, list_
        pos = len_trim( filename )
        fn_new = filename(:pos)
-       if ( masterproc ) write(iulog,*) 'incr_flnm: old filename = ',trim(fn_new)
+       if ( masterproc ) write(iulog,'(a)') 'incr_flnm: old filename = '//trim(fn_new)
```
This is more efficient as:
```suggestion
       if ( masterproc ) write(iulog,'(2a)') 'incr_flnm: old filename = ', trim(fn_new)
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -666,32 +666,32 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
            if( astat /= 0 ) then
-              write(iulog,*) 'trcdata_init: file%index0_y allocation error = ',astat
+              write(iulog,'(a,i8)') 'trcdata_init: file%index0_y allocation error = ',astat
               call endrun('trcdata_init: failed to allocate index0_y array')
```
These can be combined:
```suggestion
              call endrun('trcdata_init: failed to allocate file%index0_y array, allocation error = '//int2str(astat))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1476,14 +1476,14 @@ subroutine read_2d_trc( fid, vid, loc_arr, strt, cnt, file, order )
      if( ierr /= 0 ) then
-        write(iulog,*) 'read_2d_trc: wrk2d allocation error = ',ierr
+        write(iulog,'(a,i8)') 'read_2d_trc: wrk2d allocation error = ',ierr
         call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
        call endrun('read_2d_trc: wrk2d allocation error = '//int2str(ierr))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1476,14 +1476,14 @@ subroutine read_2d_trc( fid, vid, loc_arr, strt, cnt, file, order )
         if( ierr /= 0 ) then
-           write(iulog,*) 'read_2d_trc: wrk2d_in allocation error = ',ierr
+           write(iulog,'(a,i8)') 'read_2d_trc: wrk2d_in allocation error = ',ierr
            call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
           call endrun('read_2d_trc: wrk2d_in allocation error = '//int2str(ierr))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1591,14 +1591,14 @@ subroutine read_za_trc( fid, vid, loc_arr, strt, cnt, file, order )
      if( ierr /= 0 ) then
-        write(iulog,*) 'read_2d_trc: wrk2d allocation error = ',ierr
+        write(iulog,'(a,i8)') 'read_2d_trc: wrk2d allocation error = ',ierr
         call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
        call endrun('read_2d_trc: wrk2d allocation error = '//int2str(ierr))
```

mvertens replied:
Done

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1591,14 +1591,14 @@ subroutine read_za_trc( fid, vid, loc_arr, strt, cnt, file, order )
         if( ierr /= 0 ) then
-           write(iulog,*) 'read_2d_trc: wrk2d_in allocation error = ',ierr
+           write(iulog,'(a,i8)') 'read_2d_trc: wrk2d_in allocation error = ',ierr
            call endrun
```
Naked endrun calls are not useful (as the write statement is often lost):
```suggestion
           call endrun('read_2d_trc: wrk2d_in allocation error = '//int2str(ierr))
```
also, please fix all the `endrun` calls below.

mvertens replied:
Done

mvertens committed:
updates for namelist in aircraft_emit.F90 and removal of ocean read CO2 from file

mvertens committed:
cleanup of co2_data_flux.F90

mvertens commented:
@gold2718 - I think this is now ready for another review.

gold2718 commented on bld/build-namelist:
```diff
@@ -814,35 +814,42 @@ if ($co2_cycle) {
+            }
+            add_default($nl, 'aircraft_co2_meshfile');
+            add_default($nl, 'aircraft_co2_datafile');
+            add_default($nl, 'aircraft_co2_year_first', 'sim_year'=>$sim_year);
+            add_default($nl, 'aircraft_co2_year_last' , 'sim_year'=>$sim_year);
+            add_default($nl, 'aircraft_co2_year_align', 'sim_year'=>$sim_year);
```
Should you use the start and end years here?
```suggestion
            add_default($nl, 'aircraft_co2_year_first', 'sim_year'=>$sim_year_start);
            add_default($nl, 'aircraft_co2_year_last' , 'sim_year'=>$sim_year_end);
            add_default($nl, 'aircraft_co2_year_align', 'sim_year'=>$sim_year_start);
```

mvertens replied:
Done.
I needed to add a sim_year_end in build-namelist.
```
# sim_year_start
# If sim_year is input as a range of years, then select the first year
# to use with some datasets
my $sim_year_start = $sim_year;
my $sim_year_end = $sim_year
if ($sim_year =~ /(\d+)-(\d+)/) {
    $sim_year_start = $1;
    $sim_year_end = $2;
}
```

gold2718 commented on bld/build-namelist:
```diff
@@ -814,35 +814,42 @@ if ($co2_cycle) {
-            # This is a workaround to get this filepath into the cam.input_data_list file
-            # to allow the CESM scripts to obtain all required data for a run.
-            add_default($nl, 'aircraft_co2_file');
+            if ($sim_year =~ /(\d+)-(\d+)/) {
+                add_default($nl, 'aircraft_co2_taxmode', 'val'=>'limit' );
+            } else {
```
Do the start and end years have to be adjusted for year ranges?
```suggestion
            my $sim_year_start = $sim_year
            my $sim_year_end = $sim_year
            if ($sim_year =~ /(\d+)-(\d+)/) {
                $sim_year_start = $1
                $sim_year_end = $2
                add_default($nl, 'aircraft_co2_taxmode', 'val'=>'limit' );
            } else {
```

mvertens replied:
I did this a different way - since what you propose is partly already in place starting at line 702.

gold2718 commented on bld/build-namelist:
```diff
@@ -814,35 +814,42 @@ if ($co2_cycle) {
         # (user specification has higher precedence than the true value set above)
         if ($nl->get_value('co2_readflux_fuel') =~ /$TRUE/io) {
-            add_default($nl, 'co2flux_fuel_file', 'sim_year'=>$sim_year);
+            if ($sim_year =~ /(\d+)-(\d+)/) {
+                add_default($nl, 'co2flux_fuel_taxmode', 'val'=>'limit' );
+            } else {
```
Do the start and end years have to be adjusted for year ranges?
```suggestion
            my $sim_year_start = $sim_year
            my $sim_year_end = $sim_year
            if ($sim_year =~ /(\d+)-(\d+)/) {
                $sim_year_start = $1
                $sim_year_end = $2
                add_default($nl, 'co2flux_fuel_taxmode', 'val'=>'limit' );
            } else {
```

mvertens replied:
See above comment.

gold2718 commented on bld/build-namelist:
```diff
@@ -814,35 +814,42 @@ if ($co2_cycle) {
+            }
+            add_default($nl, 'co2flux_fuel_meshfile');
+            add_default($nl, 'co2flux_fuel_datafile');
+            add_default($nl, 'co2flux_fuel_year_first', 'sim_year'=>$sim_year);
+            add_default($nl, 'co2flux_fuel_year_last' , 'sim_year'=>$sim_year);
+            add_default($nl, 'co2flux_fuel_year_align', 'sim_year'=>$sim_year);
```
Do these need to be adjusted?
```suggestion
            add_default($nl, 'co2flux_fuel_year_first', 'sim_year'=>$sim_year_start);
            add_default($nl, 'co2flux_fuel_year_last' , 'sim_year'=>$sim_year_end);
            add_default($nl, 'co2flux_fuel_year_align', 'sim_year'=>$sim_year_start);
```

mvertens replied:
Done.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -6936,35 +6951,98 @@ Full pathname of dataset of O2 and 03 column densities above the model for look-
+<entry id="aircraft_slant_dist_fldname" type="char*256" category="cam_chem"
+       group="aircraft_emit_nl" valid_values="" >
+Field name on the aircraft_datafile (default is ac_SLANT_DIST).
+Default: set by build-namelist.
```
You have default twice and they do not agree

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -6936,35 +6951,98 @@ Full pathname of dataset of O2 and 03 column densities above the model for look-
-Filename of file that contains aircraft input file lists. The filenames in the files are relative
-to the directory specified by {{ hilight }}aircraft_datapath{{ closehilight }}.
+Field name on the aircraft_datafile (default is ac_CO2).
 Default: set by build-namelist.
```
You have default twice and they do not agree.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -6936,35 +6951,98 @@ Full pathname of dataset of O2 and 03 column densities above the model for look-
+<entry id="aircraft_h2o_fldname" type="char*256" category="cam_chem"
+       group="aircraft_emit_nl" valid_values="" >
+Field name on the aircraft_datafile (default is ac_H2O).
 Default: set by build-namelist.
```
You have default twice and they do not agree.

mvertens committed:
co2_readflux_aircraft namelist in co2_cycle.F90 is now recognoized in aircraft_emit.F90

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
+Model year that aligns with aircraft_co2_year_first.
+Default: set by build-namelist.
+</entry>
+
```
Please add `tintalgo` back to the namelist:
```suggestion

<entry id="aircraft_co2_tintalgo" type="integer" category="co2_cycle"
       group="aircraft_emit_nl" valid_values="linear, lower, upper, nearest" >
Time interpolation algorithm to use for aircraft_co2_datafile.
Default: linear
</entry>
```

mvertens replied:
Done.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
 
+<entry id="co2flux_fuel_year_first" type="integer" category="co2_cycle"
+       group="co2_ffuel_nl" valid_values="" >
+Year first to use in co2flux_fuel_datafile.
```
Grammar nitpick:
```suggestion
First year to use in co2flux_fuel_datafile.
```

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
+
+<entry id="co2flux_fuel_year_last" type="integer" category="co2_cycle"
+       group="co2_ffuel_nl" valid_values="" >
+Year last to use in co2flux_fuel_datafile.
```
Grammar nitpick
```suggestion
Last year to use in co2flux_fuel_datafile.
```

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
+
+<entry id="co2flux_fuel_year_align" type="integer" category="co2_cycle"
+       group="co2_ffuel_nl" valid_values="" >
+Year align to use in co2flux_fuel_datafile.
```
Please improve explanation:
```suggestion
Model year that aligns with co2flux_fuel_datafile_first.
```

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
+
+<entry id="co2flux_fuel_taxmode" type="char*16" category="co2_cycle"
+       group="co2_ffuel_nl" valid_values="cycle,extend,limit">
+Time extrapolation algorithm to use for co2flux_fuel_datafile.
```
Some explanation?
```suggestion
Time extrapolation algorithm to use for co2flux_fuel_datafile.
- cycle: Simply cycle through the available model data (i.e., start over when the end is reached)
- extend: Use the final value when end of data is reached.
- limit: Halt the model run if the model time exceeds the available data.
```

mvertens replied:
Good suggestion. Done.

gold2718 replied:
I do not see this change, did it not get pushed?

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
+
+<entry id="aircraft_co2_taxmode" type="char*16" category="co2_cycle"
+       group="aircraft_emit_nl" valid_values="cycle,extend,limit">
+Time extrapolation algorithm to use for aircraft_co2_datafile.
```
Add some explanation?
```suggestion
Time extrapolation algorithm to use for aircraft_co2_datafile.
- cycle: Simply cycle through the available model data (i.e., start over when the end is reached)
- extend: Use the final value when end of data is reached.
- limit: Halt the model run if the model time exceeds the available data.
```

mvertens replied:
Done.

gold2718 replied:
I would still like to see this change.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
+Year align to use in co2flux_fuel_datafile.
+Default: 0
+</entry>
+
```
Please add `co2flux_fuel_tintalgo` back to namelist:
```suggestion

<entry id="co2flux_fuel_tintalgo" type="integer" category="co2_cycle"
       group="co2_cycle_nl" valid_values="linear, lower, upper, nearest" >
Time interpolation algorithm to use for co2flux_fuel_datafile.
Default: linear
</entry>

```

mvertens replied:
Not sure what you mean by add back to namelist. It is already there for build-namelist. Can you please clarify.

gold2718 commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -659,31 +653,30 @@
-<aircraft_co2_file  hgrid="C96">atm/cam/ggas/emissions-cmip6_CO2_anthro_ac_175001-201512_fv_0.9x1.25_c20181011.nc</aircraft_co2_file>
-<aircraft_co2_file  hgrid="C48" >atm/cam/ggas/emissions-cmip6_CO2_anthro_ac_175001-201512_fv_1.9x2.5_c20181011.nc</aircraft_co2_file>
-<aircraft_co2_file  hgrid="C24" >atm/cam/ggas/emissions-cmip6_CO2_anthro_ac_175001-201512_fv_1.9x2.5_c20181011.nc</aircraft_co2_file>
+<co2flux_fuel_datafile>atm/cam/ggas/emissions-cmip6_CO2_anthro_surface_175001-201512_fv_0.9x1.25_c20181011.nc</co2flux_fuel_datafile>
```
I think it would be good to add defaults for co2_flag instead of build_namelist logic:
```suggestion
<co2_flag              > .false. </co2_flag>
<co2_flag co2_cycle="1">.true.</co2_flag>

<co2flux_fuel_datafile>atm/cam/ggas/emissions-cmip6_CO2_anthro_surface_175001-201512_fv_0.9x1.25_c20181011.nc</co2flux_fuel_datafile>
```

mvertens replied:
Done.

gold2718 commented on bld/build-namelist:
```diff
@@ -814,35 +814,42 @@ if ($co2_cycle) {
     # co2_flag turns on the co2_cycle code in CAM
     add_default($nl, 'co2_flag', 'val'=>'.true.');
```
I think this default should be moved to namelist_defaults_cam.xml (see suggestion there as well). This means moving these two lines above the `if ($co2_cycle) {` line above and removing the default:
```suggestion
# co2_flag turns on the co2_cycle code in CAM
add_default($nl, 'co2_flag');
```

mvertens replied:
Done.

gold2718 commented on bld/build-namelist:
```diff
@@ -814,35 +814,42 @@ if ($co2_cycle) {
+            add_default($nl, 'aircraft_co2_year_last' , 'sim_year'=>$sim_year);
+            add_default($nl, 'aircraft_co2_year_align', 'sim_year'=>$sim_year);
         }
     }
```
Please add defaults for the other flags to atm_in:
```suggestion
    } else {
        add_default($nl, 'co2_readflux_fuel', 'val'=>'.false.');
        add_default($nl, 'co2_readflux_aircraft', 'val'=>'.false.');
    }
```

mvertens replied:
Done.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
-If TRUE read co2 fuel flux from file.
+If TRUE read co2 aircraft flux from file and use all the settings for
+aircraft_co2_xxx namelist variables.
 Default: set by build-namelist
```
Can we explain the default here?
```suggestion
Default: True for transient model runs, false otherwise
```

mvertens replied:
@gold2718 - I don't agree with this statement. You can have co_readflux_fuel and co2_readflux_aircraft to true for a constant simulation year not just a transient run. So I'm not sure what this statement should be.

gold2718 replied:
I would still like to see this change.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -1778,36 +1778,95 @@ If TRUE turn on CO2 code.
-If TRUE read co2 aircraft flux from file.
+If TRUE read co2 fuel flux from file and use all the settings for
+co2flux_fuel_xxx namelist variables.
 Default: set by build-namelist
```
Can we explain the default here?
```suggestion
Default: True for transient model runs, false otherwise
```

mvertens replied:
See comment above.

gold2718 replied:
I would still like to see this change (assuming my logic is correct).

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      namelist /aircraft_emit_nl/  &
+           aircraft_co2_datafile, aircraft_co2_meshfile, &
+           aircraft_co2_year_first, aircraft_co2_year_last, aircraft_co2_year_align, &
+           aircraft_co2_taxmode, &
```
Please add back `tintalgo`:
```suggestion
           aircraft_co2_tintalgo, aircraft_co2_taxmode, &
```

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -6936,35 +6995,76 @@ Full pathname of dataset of O2 and 03 column densities above the model for look-
+       group="aircraft_emit_nl" valid_values="" >
+Model year that aligns with aircraft_h2o_year_first.
+</entry>
+
```
Please add `aircraft_h2o_tintalgo` back to the namelist:
```suggestion

<entry id="aircraft_h2o_tintalgo" type="integer" category="co2_cycle"
       group="co2_cycle_nl" valid_values="linear, lower, upper, nearest" >
Time interpolation algorithm to use for aircraft_h2o_datafile.
Default: linear
</entry>
```

mvertens replied:
Done.

gold2718 replied:
This is done but could you add the explanation you added to the `xxx_tintalgo` entries above?

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -6936,35 +6995,76 @@ Full pathname of dataset of O2 and 03 column densities above the model for look-
+
+<entry id="aircraft_h2o_taxmode" type="char*16" category="h2o_cycle"
+       group="aircraft_emit_nl" valid_values="cycle, extend, limit">
+Time extrapolation algorithm to use for aircraft_h2o_datafile.
```
Add some explanation?
```suggestion
Time extrapolation algorithm to use for aircraft_h2o_datafile.
- cycle: Simply cycle through the available model data (i.e., start over when the end is reached)
- extend: Use the final value when end of data is reached.
- limit: Halt the model run if the model time exceeds the available data.
```

mvertens replied:
Done.

gold2718 replied:
I would still like to see this.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -6936,35 +6995,76 @@ Full pathname of dataset of O2 and 03 column densities above the model for look-
+
+<entry id="aircraft_slant_dist_taxmode" type="char*16" category="cam_chem"
+       group="aircraft_emit_nl" valid_values="cycle,extend,limit">
+Time extrapolation algorithm to use for aircraft_slant_dist.
```
Add some explanation?
```suggestion
Time extrapolation algorithm to use for aircraft_slant_dist.
- cycle: Simply cycle through the available model data (i.e., start over when the end is reached)
- extend: Use the final value when end of data is reached.
- limit: Halt the model run if the model time exceeds the available data.
```

mvertens replied:
Done.

gold2718 replied:
I would still like to see this.

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -6936,35 +6995,76 @@ Full pathname of dataset of O2 and 03 column densities above the model for look-
+       group="aircraft_emit_nl" valid_values="" >
+Model year that aligns with aircraft_year_first.
+</entry>
+
```
Add back `tintalgo`?
```suggestion

<entry id="aircraft_slant_dist_tintalgo" type="integer" category="co2_cycle"
       group="co2_cycle_nl" valid_values="linear, lower, upper, nearest" >
Time interpolation algorithm to use for aircraft_slant_dist_datafile.
Default: linear
</entry>

```

mvertens replied:
Done.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      character(len=cs) :: aircraft_co2_fldname          = 'ac_CO2'
+      character(len=cl) :: aircraft_co2_datafile         = 'unset'
+      character(len=cl) :: aircraft_co2_meshfile         = 'unset'
+      character(len=cs) :: aircraft_co2_taxmode          = 'unset'
```
Add back `tintalgo`?
```suggestion
      character(len=cs) :: aircraft_co2_tintalgo         = 'unset'
      character(len=cs) :: aircraft_co2_taxmode          = 'unset'
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      character(len=cs) :: aircraft_h2o_fldname          = 'ac_H2O'
+      character(len=cl) :: aircraft_h2o_datafile         = 'unset'
+      character(len=cl) :: aircraft_h2o_meshfile         = 'unset'
+      character(len=cs) :: aircraft_h2o_taxmode          = 'unset'
```
Add back `tintalgo`?
```suggestion
      character(len=cs) :: aircraft_h2o_tintalgo         = 'unset'
      character(len=cs) :: aircraft_h2o_taxmode          = 'unset'
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      character(len=cs) :: aircraft_slant_dist_fldname   = 'ac_SLANT_DIST'
+      character(len=cl) :: aircraft_slant_dist_datafile  = 'unset'
+      character(len=cl) :: aircraft_slant_dist_meshfile  = 'unset'
+      character(len=cs) :: aircraft_slant_dist_taxmode   = 'unset'
```
Add back `tintalgo`?
```suggestion
      character(len=cs) :: aircraft_slant_dist_tintalgo  = 'unset'
      character(len=cs) :: aircraft_slant_dist_taxmode   = 'unset'
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+           aircraft_co2_taxmode, &
+           aircraft_h2o_datafile, aircraft_h2o_meshfile, &
+           aircraft_h2o_year_first, aircraft_h2o_year_last, aircraft_h2o_year_align, &
+           aircraft_h2o_taxmode, &
```
Please add back `tintalgo`:
```suggestion
           aircraft_h2o_tintalgo, aircraft_h2o_taxmode, &
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+           aircraft_h2o_taxmode, &
+           aircraft_slant_dist_datafile, aircraft_slant_dist_meshfile, &
+           aircraft_slant_dist_year_first, aircraft_slant_dist_year_last, aircraft_slant_dist_year_align, &
+           aircraft_slant_dist_taxmode
```
Please add back `tintalgo`:
```suggestion
           aircraft_slant_dist_tintalgo, aircraft_slant_dist_taxmode
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+               forcing(nf)%year_first = aircraft_co2_year_first
+               forcing(nf)%year_last  = aircraft_co2_year_last
+               forcing(nf)%year_align = aircraft_co2_year_align
+               forcing(nf)%taxmode    = aircraft_co2_taxmode
```
Please add back `tintalgo`:
```suggestion
               forcing(nf)%tintalgo    = aircraft_co2_tintalgo
               forcing(nf)%taxmode    = aircraft_co2_taxmode
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+            forcing(nf)%year_first = aircraft_h2o_year_first
+            forcing(nf)%year_last  = aircraft_h2o_year_last
+            forcing(nf)%year_align = aircraft_h2o_year_align
+            forcing(nf)%taxmode    = aircraft_h2o_taxmode
```
Please add back `tintalgo`:
```suggestion
            forcing(nf)%tintalgo    = aircraft_h2o_tintalgo
            forcing(nf)%taxmode    = aircraft_h2o_taxmode
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+            forcing(nf)%year_first = aircraft_slant_dist_year_first
+            forcing(nf)%year_last  = aircraft_slant_dist_year_last
+            forcing(nf)%year_align = aircraft_slant_dist_year_align
+            forcing(nf)%taxmode    = aircraft_slant_dist_taxmode
```
Please add back `tintalgo`:
```suggestion
            forcing(nf)%tintalgo    = aircraft_slant_dist_tintalgo
            forcing(nf)%taxmode    = aircraft_slant_dist_taxmode
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+         if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: forcing(nf)%year_last")
+         call mpi_bcast(forcing(nf)%year_align, 1, mpi_integer, masterprocid, mpicom, ierr)
+         if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: forcing(nf)%year_align")
+         call mpi_bcast(forcing(nf)%taxmode, len(forcing(nf)%taxmode), mpi_character, masterprocid, mpicom, ierr)
```
Please add back `tintalgo`:
```suggestion
         call mpi_bcast(forcing(nf)%tintalgo len(forcing(nf)%tintalgo), mpi_character, masterprocid, mpicom, ierr)
         call mpi_bcast(forcing(nf)%taxmode, len(forcing(nf)%taxmode), mpi_character, masterprocid, mpicom, ierr)
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+
+      ! Local variables
+      integer           :: i,idx, mm, ind, nf
+      integer           :: ierr
```
Most of these are not used, please clean up:
```suggestion
      ! Local variable
      integer           :: nf
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      !-------------------------------------------------------------------
+      use cam_history,    only: addfld, add_default
+      use phys_control,   only: phys_getopts
+      use physics_buffer, only: pbuf_get_chunk, pbuf_get_index
```
`pbuf_get_chunk` is not used and the `pbuf_get_index` call is not needed.
```suggestion
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      ! Local variables
+      type(file_desc_t) :: pioid
+      type(var_desc_t)  :: varid
+      integer           :: ierr, rc
```
I don't think `rc` is used:
```suggestion
      integer           :: ierr
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      integer           :: klev
+      integer           :: nf
+      logical           :: history_chemistry
+      character(len=3)  :: mixtype(N_AERO) = 'wet'
```
I don't think this is used:
```suggestion
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+
+      use pio, only : file_desc_t, pio_seterrorhandling
+      use pio, only : pio_inq_dimid, pio_inq_dimlen, pio_inq_varid, pio_get_var
+      use pio, only : PIO_BCAST_ERROR, PIO_NOERR
```
I think this will be useful:
```suggestion
      use pio, only : PIO_BCAST_ERROR, PIO_NOERR
      use string_utils, only : int2str
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+         allocate( data(dsize), stat=ierr )
+         if ( ierr /= 0 ) then
+            write(iulog,*) 'get_dimension: data allocation error = ',ierr
+            call endrun('get_dimension: failed to allocate data array')
```
Please fix subroutine name and include all information in the `endrun` call (in which case, the `write` statement is not necessary):
```suggestion
            write(iulog,*) 'get_dimension: data allocation error = ',ierr
            call endrun('get_vertical_dimension: failed to allocate data array, '//trim(dname)//', error = '//int2str(ierr))
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+
+   real(r8), parameter :: molmass(N_AERO) = 1._r8
+   character(*),parameter :: u_FILE_u = __FILE__
+
```
Add forward reference:
```suggestion

   private :: get_vertical_dimension

```

gold2718 replied:
Thanks and I see you also added `interpz_conserve`. Could you also add:
```suggestion
private :: chkrc
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      type(physics_buffer_desc), pointer :: pbuf2d(:,:)
+
+      ! Local variables
+      integer               :: gcell, ind, nf
```
I don't see `ind` used anywhere.
```suggestion
      integer               :: gcell, nf
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      use physconst,         only : boltz ! J/K/molecule
+      use phys_grid,         only : get_wght_all_p, get_ncols_p
+      use physics_buffer,    only : physics_buffer_desc, pbuf_get_field
+      use physics_buffer,    only : pbuf_get_chunk, pbuf_get_index
```
`pbuf_get_index` is not used:
```suggestion
      use physics_buffer,    only : pbuf_get_chunk
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      real(r8), pointer     :: tmpptr(:,:)
+      real(r8), pointer     :: data_out(:,:)
+      real(r8), pointer     :: dataptr2d(:,:)
+      real(r8)              :: datain3d(pcols,pver,begchunk:endchunk)
```
This is a lot of data to put on the stack, can you make it an allocatable?

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      character(len=cs)     :: units
+      integer               :: index
+      integer               :: rc
+      logical               :: first_time = .true.
```
This forces the `save` attribute for this routine which can cause problems in some circumstances. Can you move it to module level and maybe rename it to `first_update`?

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+      logical               :: first_time = .true.
+      type(physics_buffer_desc), pointer :: pbuf_chnk(:)
+      real(r8), parameter :: m2km  = 1.e-3_r8
+      character(len=*), parameter :: subname = 'aircraft_emit_adv'
```
This is not used. Please either remove it or use it in the `endrun` call.

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+               if (masterproc) then
+                  write(iulog,*)'aircraft_emit_adv: units = '//trim(units)//' are not recognized'
+               end if
+               call endrun('aircraft_emit_adv: units are not recognized')
```
Please include all information in the `endrun` call as the log message often does not make it to the file before the program quits.
```suggestion
               call endrun(subname//': units = '//trim(units)//' are not recognized')
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+            ! set the tracer fields with the correct units
+            !-------------------------------------------------------------------
+
+            ! GLC IS position of last significant character in string.
```
Complete nitpick:
```suggestion
            ! GLC is position of last significant character in string.
```

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,666 @@
+   end subroutine aircraft_emit_adv
+
+   !=========================================================================
+   subroutine interpz_conserve( nsrc, ndst, src_x, dst_x, src, dst)
```
It would be nice to have a comment here explaining the algorithm and/or its history.

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -254,7 +254,7 @@ subroutine trcdata_init( specifier, filename, filelist, datapath, flds, file, &
 
     if (masterproc) then
-       write(iulog,*) 'trcdata_init: data type: '//trim(data_type)//' file: '//trim(filename)
+       write(iulog,'(a)') 'trcdata_init: data type: '//trim(data_type)//' file: '//trim(filename)
```
This would be a lot more efficient as 
```suggestion
       write(iulog,'(4a)') 'trcdata_init: data type: ', trim(data_type), ' file: ', trim(filename)
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1050,8 +1050,8 @@ function incr_filename( filename, filenames_list, datapath, cyclical_list, list_
-       if ( masterproc ) write(iulog,*) 'incr_flnm: old filename = ',trim(filename)
-       if ( masterproc ) write(iulog,*) 'incr_flnm: open filenames_list : ',trim(filenames_list)
+       if ( masterproc ) write(iulog,'(a)') 'incr_flnm: old filename = '//trim(filename)
+       if ( masterproc ) write(iulog,'(a)') 'incr_flnm: open filenames_list : '//trim(filenames_list)
```
This would be more readable and efficient as:
```suggestion
       if ( masterproc ) then
          write(iulog,'(2a)') 'incr_flnm: old filename = ', trim(filename)
          write(iulog,'(2a)') 'incr_flnm: open filenames_list : ', trim(filenames_list)
       end i
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1149,7 +1149,7 @@ function incr_filename( filename, filenames_list, datapath, cyclical_list, list_
     !---------------------------------------------------------------------------------
     incr_filename = trim(fn_new)
-    if ( masterproc ) write(iulog,*) 'incr_flnm: new filename = ',trim(incr_filename)
+    if ( masterproc ) write(iulog,'(a)') 'incr_flnm: new filename = '//trim(incr_filename)
```
This would be more efficient as
```suggestion
    if ( masterproc ) write(iulog,'(2a)') 'incr_flnm: new filename = ', trim(incr_filename)
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1238,7 +1238,7 @@ subroutine find_times( itms, fids, time, file, datatimem, datatimep, times_found
        if (masterproc) then
           write(iulog,*)'FIND_TIMES: Failed to find dates bracketing desired time =', time
-          write(iulog,*) 'filename = '//trim(file%curr_filename)
+          write(iulog,'(a)') 'filename = '//trim(file%curr_filename)
```
This would be more efficient as:
```suggestion
          write(iulog,'(2a)') 'filename = ', trim(file%curr_filename)
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1247,7 +1247,7 @@ subroutine find_times( itms, fids, time, file, datatimem, datatimep, times_found
     if( astat/= 0 ) then
-       write(iulog,*) 'find_times: failed to deallocate all_data_times array; error = ',astat
+       write(iulog,'(a,i8)') 'find_times: failed to deallocate all_data_times array; error = ',astat
        call endrun
```
Please eliminate all naked `endrun` calls:
```suggestion
       call endrun('find_times: failed to deallocate all_data_times array; error = `'//int2str(astat`))
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1731,7 +1731,7 @@ subroutine read_3d_trc( fid, vid, loc_arr, strt, cnt, file, order)
     if( ierr /= 0 ) then
-       write(iulog,*) 'read_3d_trc: wrk3d allocation error = ',ierr
+       write(iulog,'(a,i8)') 'read_3d_trc: wrk3d allocation error = ',ierr
        call endrun
```
Please eliminate naked `endrun` calls:
```suggestion
       call endrun('read_3d_trc: wrk3d allocation error = '//int2str(ierr))
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1741,7 +1741,7 @@ subroutine read_3d_trc( fid, vid, loc_arr, strt, cnt, file, order)
        if( ierr /= 0 ) then
-          write(iulog,*) 'read_3d_trc: wrk3d allocation error = ',ierr
+          write(iulog,'(a,i8)') 'read_3d_trc: wrk3d allocation error = ',ierr
           call endrun
```
Please eliminate naked `endrun` calls (here and below, there are several more):
```suggestion
          call endrun('read_3d_trc: wrk3d allocation error = '//int2str(ierr))
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -2129,31 +2129,31 @@ subroutine open_trc_datafile( fname, path, piofile, times, cyc_ndx_beg, cyc_ndx_
     call getfil( filepath, filen, 0 )
     call cam_pio_openfile( piofile, filen, PIO_NOWRITE)
-    if(masterproc) write(iulog,*)'open_trc_datafile: ',trim(filen)
+    if(masterproc) write(iulog,'(a)')'open_trc_datafile: '//trim(filen)
```
This would be more efficient as:
```suggestion
    if(masterproc) write(iulog,'(2a)')'open_trc_datafile: ', trim(filen)
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -1050,8 +1050,8 @@ function incr_filename( filename, filenames_list, datapath, cyclical_list, list_
-       if ( masterproc ) write(iulog,*) 'incr_flnm: open filenames_list : ',trim(filenames_list)
+       if ( masterproc ) write(iulog,'(a)') 'incr_flnm: old filename = '//trim(filename)
+       if ( masterproc ) write(iulog,'(a)') 'incr_flnm: open filenames_list : '//trim(filenames_list)
        unitnumber = shr_file_getUnit()
```
This should be removed along with the use statement above and the `shr_file_freeUnit` call below. In addition, change the `open` calls below to use `newunit=unitnumber`.
```suggestion
```

gold2718 replied:
I would still like to see this change

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -2191,18 +2191,18 @@ subroutine open_trc_datafile( fname, path, piofile, times, cyc_ndx_beg, cyc_ndx_
        if (cyc_ndx_beg < 0) then
-          write(iulog,*) 'open_trc_datafile: cycle year not found : ' , cyc_yr
+          write(iulog,'(a,i8)') 'open_trc_datafile: cycle year not found : ' , cyc_yr
           call endrun('open_trc_datafile: cycle year not found '//trim(filepath))
```
Please include all information in the `endrun` call:
```suggestion
          call endrun('open_trc_datafile: cycle year, '//int2str(cyc_yr)//', not found in '//trim(filepath))
```

gold2718 commented on src/chemistry/utils/tracer_data.F90:
```diff
@@ -2804,10 +2804,10 @@ subroutine advance_file(file)
     if( file%remove_trc_file ) then
        call getfil( file%curr_filename, loc_fname, 0 )
-       write(iulog,*) 'advance_file: removing file = ',trim(loc_fname)
+       write(iulog,'(a)') 'advance_file: removing file = ',trim(loc_fname)
```
This `write` statement has 2 output fields:
```suggestion
       write(iulog,'(2a)') 'advance_file: removing file = ',trim(loc_fname)
```

gold2718 commented on src/control/cam_esmf_mod.F90:
```diff
@@ -0,0 +1,167 @@
+module cam_esmf_mod
+
+  use shr_kind_mod      , only : r8=>shr_kind_r8, i8=>shr_kind_i8, cl=>shr_kind_cl, cs=>shr_kind_cs
```
I think most of these are not used:
```suggestion
  use shr_kind_mod      , only : r8=>shr_kind_r8
```

gold2718 commented on src/control/cam_esmf_mod.F90:
```diff
@@ -0,0 +1,167 @@
+  use ESMF              , only : ESMF_VM, ESMF_VMAllreduce, ESMF_VMGetCurrent
+  use ESMF              , only : ESMF_SUCCESS, ESMF_REDUCE_SUM
+  use shr_sys_mod       , only : shr_sys_abort
+  use cam_abortutils    , only : endrun
```
I don't think this is called (but `shr_sys_abort` is):
```suggestion
```

mvertens replied:
Now calling endrun and not shr_sys_abort.

gold2718 commented on src/control/cam_esmf_mod.F90:
```diff
@@ -0,0 +1,167 @@
+      else
+         model_clock = clock_in
+         model_clock_initialized = .true.
+      end if
```
It feels a bit weird to check the initialized flag _after_ you have already created a clock. Also, I wonder if `model_clock_initialized` is even needed:
```suggestion

      if (ESMF_ClockIsCreated(model_clock, rc=rc)) then
         call shr_sys_abort('initialize_model_clock: model clock already initialized')
      else
         model_clock = ESMF_ClockCreate(clock_in, rc=rc)
         if (ChkErr(rc,__LINE__,u_FILE_u)) return
         model_clock = clock_in
      end if
```

mvertens replied:
I've removed the initialized checks from model_clock and model_mesh.

gold2718 commented on src/control/cam_esmf_mod.F90:
```diff
@@ -0,0 +1,167 @@
+      if (ChkErr(rc,__LINE__,u_FILE_u)) return
+
+      write(iulog,'(a,d13.5)') ' global mesh area  = ',global_mesh_area(1)
+      write(iulog,'(a,d13.5)') ' global model area = ',global_model_area(1)
```
Should these only be printed on `masterproc` (which needs to be imported)?
```suggestion
      if (masterproc) then
         write(iulog,'(a,d13.5)') ' global mesh area  = ',global_mesh_area(1)
         write(iulog,'(a,d13.5)') ' global model area = ',global_model_area(1)
      end if
```

mvertens replied:
This has been refactored and the write statements are no longer in this routine.

gold2718 commented on src/control/cam_esmf_mod.F90:
```diff
@@ -0,0 +1,167 @@
+
+      write(iulog,'(a)') 'Global sum for forcing field '//trim(fldname)
+      write(iulog,'(a,d13.5)') ' global sum with model areas = ',global_sum_model(1)
+      write(iulog,'(a,d13.5)') ' global sum with mesh areas  = ',global_sum_mesh(1)
```
Should these only be printed on `masterproc`?
```suggestion
      if (masterproc) then
         write(iulog,'(a)') 'Global sum for forcing field '//trim(fldname)
         write(iulog,'(a,d13.5)') ' global sum with model areas = ',global_sum_model(1)
         write(iulog,'(a,d13.5)') ' global sum with mesh areas  = ',global_sum_mesh(1)
      end if
```

gold2718 commented on src/control/cam_esmf_mod.F90:
```diff
@@ -0,0 +1,167 @@
+  logical :: model_clock_initialized = .false.
+  logical :: model_mesh_initialized = .false.
+
+  character(*), parameter :: u_FILE_u = &
```
Please correct syntax:
```suggestion
  character(len=*), parameter :: u_FILE_u = &
```

mvertens replied:
Done.

gold2718 commented on src/cpl/nuopc/atm_comp_nuopc.F90:
```diff
@@ -11,6 +11,7 @@ module atm_comp_nuopc
    use ESMF                , only : ESMF_DistGrid, ESMF_DistGridCreate
    use ESMF                , only : ESMF_Mesh, ESMF_MeshCreate, ESMF_MeshGet, ESMF_FILEFORMAT_ESMFMESH
    use ESMF                , only : ESMF_Clock, ESMF_ClockGet, ESMF_ClockSet, ESMF_ClockGetNextTime, ESMF_ClockAdvance
+   use ESMF                , only : ESMF_CLockCreate
```
This is no longer used, please remove
```suggestion
```

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -1,22 +1,20 @@
+   !-------------------------------------------------------------------------------
 
-   implicit none
+   use shr_kind_mod,  only: r8=>shr_kind_r8, cl=>shr_kind_cl, cs=>shr_kind_cs
```
I don't think the character lengths are used:
```suggestion
   use shr_kind_mod,  only: r8=>shr_kind_r8
```

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
-      c_names = (/'CO2_OCN', 'CO2_FFF', 'CO2_LND', 'CO2    '/)
+        c_names = (/'CO2_OCN', 'CO2_FFF', 'CO2_LND', 'CO2    '/)
 
    integer :: co2_ocn_glo_ind ! global index of 'CO2_OCN'
```
Is this used anywhere? Maybe remove it along with the code that sets it below?

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
 
    integer :: co2_ocn_glo_ind ! global index of 'CO2_OCN'
    integer :: co2_fff_glo_ind ! global index of 'CO2_FFF'
    integer :: co2_lnd_glo_ind ! global index of 'CO2_LND'
```
Is this used anywhere? Maybe remove it along with the code that sets it below?

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
-   end if
+         select case (trim(c_names(icnst)))
+         case ('CO2_OCN')
+            co2_ocn_glo_ind = c_i(icnst)
```
Is this used for anything? Can it be removed?

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
+         case ('CO2_FFF')
+            co2_fff_glo_ind = c_i(icnst)
+         case ('CO2_LND')
+            co2_lnd_glo_ind = c_i(icnst)
```
Is this used for anything? Can it be removed?

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -73,15 +73,15 @@ subroutine getfil(fulpath, locfn, iflag, lexist)
          call endrun('(GETFIL): local filename variable is too short for path length')
       else
-         if (masterproc) write(iulog,*) '(GETFIL): local filename variable is too short for path length',klen-i,maxlen
+         if (masterproc) write(iulog,'(a,i8,i8)') '(GETFIL): local filename variable is too short for path length',klen-i,maxlen
```
As long as we are trying to improve messages:
```suggestion
         if (masterproc) write(iulog,'(a,i0,a,i0)') '(GETFIL): local filename variable is too short for path length: ',klen-i,' > ',maxlen
```

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -91,23 +91,23 @@ subroutine getfil(fulpath, locfn, iflag, lexist)
       call endrun ('(GETFIL): local filename has zero length')
    else if (masterproc) then
-      write(iulog,*)'(GETFIL): attempting to find local file ', trim(locfn)
+      write(iulog,'(a)')'(GETFIL): attempting to find local file '//trim(locfn)
```
This would be more efficient as:
```suggestion
      write(iulog,'(2a)')'(GETFIL): attempting to find local file ', trim(locfn)
```

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -91,23 +91,23 @@ subroutine getfil(fulpath, locfn, iflag, lexist)
    if (present(lexist)) lexist = lexist_in
    if (lexist_in) then
-      if (masterproc) write(iulog,*) '(GETFIL): using ',trim(locfn), ' in current working directory'
+      if (masterproc) write(iulog,'(a)') '(GETFIL): using '//trim(locfn)//' in current working directory'
```
This would be more efficient as:
```suggestion
      if (masterproc) write(iulog,'(3a)') '(GETFIL): using ', trim(locfn), ' in current working directory'
```

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -91,23 +91,23 @@ subroutine getfil(fulpath, locfn, iflag, lexist)
          call endrun('(GETFIL): local filename variable is too short for path length')
       else
-         if (masterproc) write(iulog,*) '(GETFIL): local filename variable is too short for path length',klen,maxlen
+         if (masterproc) write(iulog,'(a,i8,i8)') '(GETFIL): local filename variable is too short for path length',klen,maxlen
```
As long as we are improving write messages:
```suggestion
         if (masterproc) write(iulog,'(a,i0,a,i0)') '(GETFIL): local filename variable is too short for path length: ',klen,' > ',maxlen
```

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -117,10 +117,10 @@ subroutine getfil(fulpath, locfn, iflag, lexist)
    if (present(lexist)) lexist = lexist_in
    if (lexist_in) then
-      if (masterproc) write(iulog,*)'(GETFIL): using ',trim(fulpath)
+      if (masterproc) write(iulog,'(a)')'(GETFIL): using '//trim(fulpath)
```
This would be more efficient as:
```suggestion
      if (masterproc) write(iulog,'(2a)')'(GETFIL): using ', trim(fulpath)
```

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -117,10 +117,10 @@ subroutine getfil(fulpath, locfn, iflag, lexist)
       return
    else
-      if (masterproc) write(iulog,*)'(GETFIL): all tries to get file have been unsuccessful: ',trim(fulpath)
+      if (masterproc) write(iulog,'(a)')'(GETFIL): all tries to get file have been unsuccessful: '//trim(fulpath)
```
This would be more efficient as:
```suggestion
      if (masterproc) write(iulog,'(2a)')'(GETFIL): all tries to get file have been unsuccessful: ', trim(fulpath)
```

mvertens replied:
Done.

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -167,16 +167,16 @@ subroutine opnfil (locfn, iun, form, status)
    if (ioe /= 0) then
-      if(masterproc) write(iulog,*)'(OPNFIL): failed to open file ',trim(locfn), ' on unit ',iun,' ierr=',ioe
-      call endrun ('opnfil') 
+      if(masterproc) write(iulog,'(a,i8,a,i8)')'(OPNFIL): failed to open file '//trim(locfn)//' on unit ',iun,' ierr=',ioe
```
As long as we are improving write messages:
```suggestion
      if(masterproc) write(iulog,'(3a,i0,a,i0)')'(OPNFIL): failed to open file ', trim(locfn), ' on unit ',iun,',  ierr=',ioe
```

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -167,16 +167,16 @@ subroutine opnfil (locfn, iun, form, status)
-      if(masterproc) write(iulog,*)'(OPNFIL): failed to open file ',trim(locfn), ' on unit ',iun,' ierr=',ioe
-      call endrun ('opnfil') 
+      if(masterproc) write(iulog,'(a,i8,a,i8)')'(OPNFIL): failed to open file '//trim(locfn)//' on unit ',iun,' ierr=',ioe
+      call endrun ('opnfil')
```
Please include all information in the `endrun` call:
```suggestion
      call endrun ('(OPNFIL): failed to open file '//trim(locfn)//' on unit '//int2str(iun)//',  ierr='//int2str(ioe))
```

gold2718 commented on src/utils/ioFileMod.F90:
```diff
@@ -167,16 +167,16 @@ subroutine opnfil (locfn, iun, form, status)
+      call endrun ('opnfil')
    else
-      if(masterproc) write(iulog,*)'(OPNFIL): Successfully opened file ',trim(locfn), ' on unit= ',iun
+      if(masterproc) write(iulog,'(a,i8)')'(OPNFIL): Successfully opened file '//trim(locfn)//' on unit= ',iun
```
As long as we are improving write messages:
```suggestion
      if(masterproc) write(iulog,'(3a,i0)')'(OPNFIL): Successfully opened file ', trim(locfn), ' on unit = ', iun
```

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
-   end do
-
-end subroutine co2_init_cnst
+   end subroutine co2_cycle_readnl
```
Can you please output the namelist values to the log?

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
+   !===============================================================================
 
-!===============================================================================
+   function co2_transport()
```
Please be specific
```suggestion
   logical function co2_transport()
```

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
+   !===============================================================================
 
-end subroutine co2_time_interp_ocn
+   function co2_implements_cnst(name)
```
Specify type:
```suggestion
   logical function co2_implements_cnst(name)
```

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
+      real(r8),         intent(out) :: q(:,:)     ! kg tracer/kg dry air (gcol, plev)
+
+      ! Local variables
+      integer :: k
```
I'm not a big fan of single-letter variables. If you are up to it, please improve.

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
-      ptend%q(:ncol,k,co2_fff_glo_ind) = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
-      ptend%q(:ncol,k,co2_glo_ind)     = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
-   end do
+      ifld = pbuf_get_index('ac_CO2')
```
`ifld` should be a module variable with a more specific name (e.g., `idx_ac_CO2`) and initialized to `-1`. The `pbuf_get_index` call should be moved to `co2_init`.

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
-      ptend%q(:ncol,k,co2_glo_ind)     = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
-   end do
+      ifld = pbuf_get_index('ac_CO2')
+      call pbuf_get_field(pbuf, ifld, ac_CO2)
```
This code (and the code below) should only be called if the field exists:
```suggestion
      if (idx_ac_CO2 > 0) then
         call pbuf_get_field(pbuf, ifld, ac_CO2)
```

gold2718 commented on src/physics/cam/co2_cycle.F90:
```diff
@@ -26,402 +24,312 @@ module co2_cycle
+      do k = 1, pver
+         ptend%q(:ncol,k,co2_fff_glo_ind) = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
+         ptend%q(:ncol,k,co2_glo_ind)     = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
+      end do
```
see comment above:
```suggestion
         ! [ac_CO2] = 'kg m-2 s-1'
         ! [ptend%q] = 'kg kg-1 s-1'
         ncol = state%ncol
         do k = 1, pver
            ptend%q(:ncol,k,co2_fff_glo_ind) = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
            ptend%q(:ncol,k,co2_glo_ind)     = gravit * state%rpdeldry(:ncol,k) * ac_CO2(:ncol,k)
         end do
      end if
```

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+      call mpi_bcast(co2flux_fuel_taxmode, len(co2flux_fuel_taxmode), mpi_character, masterprocid, mpicom, ierr)
+      if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: co2flux_fuel_taxmode")
+
+   end subroutine co2_data_flux_readnl
```
Please output the namelist values to the `iulog`.

gold2718 replied:
I would still like to see these values in the log

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
-   xin%initialized = .true.
-
-end subroutine co2_data_flux_init
+   character(*),parameter :: u_FILE_u = __FILE__
```
Please fix syntax:
```suggestion
   character(len=*),parameter :: u_FILE_u = __FILE__
```

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+
+      ! Local variables
+      integer            :: unitn, ierr
+      character(len=256) :: msg
```
This is not being used, please remove.
```suggestion
```

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+           co2flux_fuel_year_first, & ! first year in stream to use
+           co2flux_fuel_year_last,  & ! last year in stream to use
+           co2flux_fuel_year_align, & ! align stream_year_first
+           co2flux_fuel_taxmode       ! time extraploation [cycle, extend or limit]
```
If we are adding `tintalgo` back to the namelist:
```suggestion
           co2flux_fuel_tintalgo      ! time interpolation [linear, lower, upper, nearest]

           co2flux_fuel_taxmode       ! time extraploation [cycle, extend or limit]
```

gold2718 replied:
The comment has the wrong word on line 84 `extrapolation ==> interpolation`.

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+      call mpi_bcast(co2flux_fuel_year_align, 1, mpi_integer, masterprocid, mpicom, ierr)
+      if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: co2flux_fuel_year_align")
+      call mpi_bcast(co2flux_fuel_taxmode, len(co2flux_fuel_taxmode), mpi_character, masterprocid, mpicom, ierr)
+      if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: co2flux_fuel_taxmode")
```
If we are adding `tintalgo` back to the namelist:
```suggestion
      call mpi_bcast(co2flux_fuel_tintalgo, len(co2flux_fuel_tintalgo), mpi_character, masterprocid, mpicom, ierr)
      if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: co2flux_fuel_tintalgo")
      call mpi_bcast(co2flux_fuel_taxmode, len(co2flux_fuel_taxmode), mpi_character, masterprocid, mpicom, ierr)
      if (ierr /= 0) call endrun(subname//": FATAL: mpi_bcast: co2flux_fuel_taxmode")
```

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+   integer            :: co2flux_fuel_year_first = -999  ! first year in stream to use
+   integer            :: co2flux_fuel_year_last = -999   ! last year in stream to use
+   integer            :: co2flux_fuel_year_align = -999  ! align stream_year_first
+   character(len=cs)  :: co2flux_fuel_taxmode = 'unset'  ! time extraploation [cycle, extend or limit]
```
If we are adding `tintalgo` back to the namelist:
```suggestion
   character(len=cs)  :: co2flux_fuel_tintalgo = 'unset'  ! time interpolation algorithm [linear, lower, upper, nearest]
   character(len=cs)  :: co2flux_fuel_taxmode = 'unset'  ! time extraploation [cycle, extend or limit]
```

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+           stream_offset       = 0,                           &
+           stream_taxmode      = trim(co2flux_fuel_taxmode),  &
+           stream_dtlimit      = 1.0e30_r8,                   &
+           stream_tintalgo     = 'linear',                    &
```
If we are adding `tintalgo` back to the namelist:
```suggestion
           stream_tintalgo     = trim(co2flux_fuel_tintalgo),  &
```

mvertens replied:
Done.

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+      use cam_esmf_mod     , only : cam_esmf_global_sum
+
+      ! Local variables
+      integer :: icol,lchnk,g
```
Can `g` please be turned into a multi-character variable (e.g., `gidx`)?

mvertens replied:
Done.

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+      integer :: sec     ! seconds into current date for nstep+1
+      integer :: mcdate  ! Current model date (yyyymmdd)
+      integer :: rc
+      logical :: first_time = .true.
```
This silently adds the `save` attribute to the subroutine which we do not want. Please move this to the module scope (e.g., as `first_advance_call`).

mvertens replied:
Good point. Done.

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+      integer :: rc
+      logical :: first_time = .true.
+      real(r8), pointer :: dataptr1d(:)
+      character(len=*), parameter :: subname = 'co2_data_flux_advance'
```
This is not used, remove?

mvertens replied:
Done.

gold2718 commented on src/physics/cam/co2_data_flux.F90:
```diff
@@ -1,164 +1,225 @@
+
+      if ( rc /= ESMF_SUCCESS ) then
+         call ESMF_LogWrite('ERROR:', ESMF_LOGMSG_ERROR, line=line, file=file)
+         call endrun('chkrc')
```
It would be nice to point users to the ESMF log files here. Something like:
```suggestion
         call endrun('chkrc: See ESMF log files (filenames begin with PET)')
```

mvertens replied:
Done.

gold2718 reviewed (changes requested):
Thanks for this huge effort. I think we need another round of cleanup and many of my comments relate to restoring `tintalgo` to the namelist for various emission types.

mvertens committed:
made tintalgo a namelist variable

mvertens committed:
addressed issues in PR

mvertens committed:
more changes for PR comments

mvertens committed:
added documentation requested in PR

gold2718 commented on src/chemistry/utils/aircraft_emit.F90:
```diff
@@ -1,458 +1,709 @@
+               endif
+            end do
+            if ( index < 1 ) then
+               call endrun('aircraft_emit_register: '//trim(forcing(nf)%fldname)//&
```
This is the wrong subroutine name:
```suggestion
               call endrun(subname//": "//trim(forcing(nf)%fldname)//&
```

mvertens committed:
addressed more issues raised in the PR review

mvertens committed:
Update src/control/cam_esmf_mod.F90

Co-authored-by: goldy <1588651+gold2718@users.noreply.github.com>

gold2718 reviewed (changes requested):
Thanks, that was a lot of clean. There are still some items on the original list plus maybe one new item. I know that GitHub is making it hard to see some of the comments so I recommented on several.

mvertens committed:
addressed more comments in PR

mvertens committed:
yet more changes to address PR comments

mvertens commented:
@gold2718 - I think I've addressed all the changes you requested - but it was not always possible for me to respond to the suggestions. I'm happy to address any issues that have not yet been resolved.

Commit:
Addressed final PR comments and requests

Commit:
Try fixing clubb submodule hash

Commit:
Merge remote-tracking branch 'NorESM/noresm_develop' into feature/refactor_co2input

Commit:
Fix compiler errors

mvertens committed:
fixed problems in namelist generation

Commit:
Fix typos in build-namelist

Commit:
Fix use of sim_year and add to NorESM use-case files

Commit:
Correct some namelist documentation

gold2718 reviewed (approved):
I think this is finally ready to go, yay!

gold2718 commented:
Tagged as `noresm3_0_022_cam6_4_121`
